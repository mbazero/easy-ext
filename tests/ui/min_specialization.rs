#![feature(min_specialization)]

// See also run-pass/min_specialization.rs.

// https://github.com/rust-lang/rust/blob/7bd62a8f5a4d6d740677aea03c37771258529922/src/test/ui/specialization/min_specialization/spec-iter.rs
pub mod spec_iter {
    use easy_ext::ext;

    #[ext(SpecFromIter)]
    impl<'a, T: 'a, I: Iterator<Item = &'a T>> I {
        default fn f(&self) {}
    }

    // This will not compile because all of the generics in the default
    // implementation are used in the generics of the trait.
    // To fix compile error, you will need to add `'a` to `SpecFromIter`.
    // See also spec_iter module in run-pass/min_specialization.rs.
    impl<'a, T> SpecFromIter<T> for std::slice::Iter<'a, T> {
        fn f(&self) {}
    }
}

fn main() {}
