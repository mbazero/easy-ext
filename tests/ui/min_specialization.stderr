error[E0726]: implicit elided lifetime not allowed here
  --> $DIR/min_specialization.rs:18:17
   |
18 |     impl<'a, T> SpecFromIter<T> for std::slice::Iter<'a, T> {
   |                 ^^^^^^^^^^^^^^^ help: indicate the anonymous lifetime: `SpecFromIter<'_, T>`

error[E0495]: cannot infer an appropriate lifetime for lifetime parameter `'a` due to conflicting requirements
   |
note: first, the lifetime cannot outlive the lifetime `'a` as defined on the impl at 18:10...
  --> $DIR/min_specialization.rs:18:10
   |
18 |     impl<'a, T> SpecFromIter<T> for std::slice::Iter<'a, T> {
   |          ^^
   = note: ...so that the types are compatible
   = note: expected `Iterator`
              found `Iterator`
   = note: but, the lifetime must be valid for the static lifetime...
   = note: ...so that the types are compatible
   = note: expected `SpecFromIter<'static, T>`
              found `SpecFromIter<'_, T>`
